binary_search_tree/tree.py

```diff
@@ -14,42 +14,120 @@ class Tree:
+        if current_node is None:
+            return new_node
+
+        elif new_node.key < current_node.key:
+            current_node.left = self.add_helper(current_node.left, new_node)
+        else:
+            current_node.right = self.add_helper(current_node.right, new_node)
+        return current_node
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1) 
     def add(self, key, value = None):
-        pass
+        new_node = TreeNode(key, value)
+
+        if self.root is None:
+            self.root = new_node
+            return self.root
+        self.add_helper(self.root, new_node)
+
+    def find_helper(self, current_node, key):
+        if current_node == None:
+            return None
+        elif current_node.key == key:
+            return current_node.value
+        elif current_node.key > key:
+            return self.find_helper(current_node.left, key)
+        else:
+            return self.find_helper(current_node.right, key)
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1) 
     def find(self, key):
-        pass
+        if self.root == None:
+            return None
 
-    # Time Complexity: 
-    # Space Complexity: 
+        return self.find_helper(self.root, key)
+        
+
+    def inorder_helper(self, current_node, items):
+        if current_node is not None:
+            self.inorder_helper(current_node.left, items)
+            items.append({"key": current_node.key, "value":current_node.value})
+            self.inorder_helper(current_node.right, items)
+
+    # Time Complexity:O(n) 
+    # Space Complexity: O(n)
     def inorder(self):
-        pass
+        items = []
+
+        self.inorder_helper(self.root, items)
+        return items
 
-    # Time Complexity: 
-    # Space Complexity:     
+    def preorder_helper(self, current_node, items):
+        if current_node is not None:
+            items.append({"key": current_node.key, "value":current_node.value})
+            self.preorder_helper(current_node.left, items)
+            self.preorder_helper(current_node.right, items)
+        
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)  
     def preorder(self):
-        pass
+        items = []
+        self.preorder_helper(self.root, items)
+        return items
+
+
+    def postorder_helper(self, current_node, items):
+        if current_node is not None:
+            self.postorder_helper(current_node.left, items)
+            self.postorder_helper(current_node.right, items)
+            items.append({"key":current_node.key, "value": current_node.value})
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)  
     def postorder(self):
-        pass
+        items = []
+        self.postorder_helper(self.root, items)
+        return items
+
+    # Time Complexity:O(n) 
+    # Space Complexity: O(n) 
+
+    def height_helper(self, current_node):
```

Comment:
✨Nice! Time and space complexity would be O(n) if the tree is unbalanced! For a balanced tree, time and space complexity would be O(log n) because your recursive calls only go through half the nodes!